src/main/java/com/varukha/onlinebookstore/service/shoppingCart/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,105 @@
+import lombok.RequiredArgsConstructor;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemMapper cartItemMapper;
+    private final CartItemRepository cartItemRepository;
+    private final BookRepository bookRepository;
+
+    @Override
+    public void create(User user) {
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCart.setUser(user);
+        shoppingCartRepository.save(shoppingCart);
+    }
+
+    @Transactional
+    @Override
+    public CartItemDto save(CreateCartItemRequestDto requestDto) {
+        CartItem existedCartItem = getCartItemByBookId(requestDto.getBookId());
+        if (existedCartItem == null || !existedCartItem.getBook().getId()
+                .equals(requestDto.getBookId())) {
+            return createNewCartItem(requestDto);
+        } else {
+            return updateCartItemQuantityWhenSave(existedCartItem, requestDto.getQuantity());
+        }
+    }
+
+    @Override
+    public CartItem getCartItemById(Long id) {
+        return cartItemRepository.findById(id).orElse(null);
+    }
+
+    @Override
+    public ShoppingCartDto getShoppingCartByUserId(Long id) {
+        return shoppingCartRepository.findByUserId(getUserId())
+                .map(shoppingCartMapper::toDto)
+                .orElseThrow();
+    }
+
+    @Override
+    public void deleteById(Long id) {
+        cartItemRepository.deleteById(id);
+    }
+
+    @Transactional
+    @Override
+    public CartItemDto updateCartItemQuantity(Long id,
+                                              UpdateCartItemRequestDto cartItemRequestDto) {
+        CartItem cartItemById = getCartItemById(id);
+        int requestedQuantity = cartItemRequestDto.getQuantity();
+        cartItemById.setQuantity(requestedQuantity);
+        CartItem savedCartItem = cartItemRepository.save(cartItemById);
+        return cartItemMapper.toDto(savedCartItem);
+    }
+
+    private CartItemDto updateCartItemQuantityWhenSave(CartItem existingCartItem, int quantity) {
+        existingCartItem.setQuantity(existingCartItem.getQuantity() + quantity);
+        CartItem savedCartItem = cartItemRepository.save(existingCartItem);
+        return cartItemMapper.toDto(savedCartItem);
+    }
+
+    private CartItemDto createNewCartItem(CreateCartItemRequestDto requestDto) {
+        CartItem cartItem = new CartItem();
+        cartItem.setShoppingCart(shoppingCartRepository
+                .findByUserId(getUserId())
+                .orElseThrow());
```

Comment:
Add information here

Reply:
What information do you mean?

Reply:
Something like {“Can't find shopping cart by user id” + id}